syntax = "proto3";

package text_service;
option csharp_namespace = "Everywhere";

import "google/protobuf/empty.proto";

message Rect {
  int32 left = 1;
  int32 top = 2;
  int32 right = 3;
  int32 bottom = 4;
}

message InitializedMessage {
  uint32 pid = 1;
}

message FocusChangedMessage {
  uint32 pid = 1;
  uint64 ctx = 2;
  uint64 hwnd = 3;
  uint64 prev_ctx = 4;
  uint64 prev_hwnd = 5;
  Rect screen_rect = 6;
}

message FocusTextMessage {
  string text = 1;
}

message EndEditMessage {
  uint32 pid = 1;
  uint64 ctx = 2;
  uint64 wnd = 3;
}

message ClientMessage {
  oneof data {
    InitializedMessage initialized = 1;
    FocusChangedMessage focus_changed = 2;
    FocusTextMessage focus_text = 3;
    EndEditMessage end_edit = 4;
  }
}

message GetFocusText {
  bool selection_only = 1;
}

message SetFocusText {
  string text = 1;
  bool append = 2;
}

message ServerMessage {
  oneof data {
    GetFocusText get_focus_text = 1;
    SetFocusText set_focus_text = 2;
  }
}